In [1]:
import pandas as pd
import re
import yaml
from data_io import load_data_from_db, insert_collection_into_db
from merge import merge_by_key_column
from tokenize import tokenize, NAME
from io import BytesIO
import warnings
warnings.filterwarnings("ignore")

In [2]:
pd.set_option('max_colwidth', 100)
pd.set_option('display.max_columns', None)

In [3]:
with open('params.yml', 'r') as f:
    params = yaml.full_load(f)

class_name = 'Bay'

df_rs, df_supa, df_voltages_matched = load_data_from_db(class_name, params)

In [4]:
for df in [df_rs, df_supa, df_voltages_matched]:
    display(df.head(10))
    display(df.info())
    print('-'*10)

_id                                   Uid  \
0  65a4dad6e20c332a8f154f33  c1413450-2c4e-4505-980b-ed4cc09737e4   
1  65a4dad6e20c332a8f154f34  f9735b10-b60d-446a-b539-08b645e8a533   
2  65a4dad6e20c332a8f154f35  cda400a2-f2db-4796-bb80-7bc2a9d652be   
3  65a4dad6e20c332a8f154f36  167d5423-42aa-458e-90f9-4dc67013adea   
4  65a4dad6e20c332a8f154f37  6a0433c6-f9b4-4d42-bb5e-4350177582e8   
5  65a4dad6e20c332a8f154f38  0ca08c1c-95f1-4ea0-841a-c83d81822e37   
6  65a4dad6e20c332a8f154f39  a066cebe-a5a0-4af7-83cd-c34cc9641113   
7  65a4dad6e20c332a8f154f3a  4fa1d943-efc2-4084-a3a6-41e6eb1bd482   
8  65a4dad6e20c332a8f154f3b  eff86a46-f961-4a7a-ab94-30ee17410d5d   
9  65a4dad6e20c332a8f154f3c  8b8de086-b54e-4527-a238-79ff92a14d06   

          name                                Класс  \
0       РП-110  Группа коммутационного оборудования   
1       РП-110  Группа коммутационного оборудования   
2       СП-110  Группа коммутационного оборудования   
3       СП-110  Группа коммутационного оборудования   
4       СП-110  Группа коммутационного оборудования   
5          Т-1  Группа коммутационного оборудования   
6  ЗОН-110 Т-1  Группа коммутационного оборудования   
7      ЗОН Т-1  Группа коммутационного оборудования   
8  ЗОН-110 Т-2  Группа коммутационного оборудования   
9      ЗОН Т-2  Группа коммутационного оборудования   

                                Bay_Uid  
0  f882ca3c-1f15-43ee-b604-834e430e3a2e  
1  4d87d79c-3265-4689-8e93-aac597307e63  
2  063e4acc-c24b-4ab6-b63a-e65bdc84f088  
3  0ae198e5-68dd-409d-ad3a-c1ac6304002d  
4  7bb39d2d-d23f-4c10-8807-34220c086532  
5  651206c7-a029-45b7-926f-47448c0510a8  
6  c30d0232-10ee-45f0-9005-9270512c4abc  
7  fe9b7170-8dc5-47e5-944a-93a8d1812b9f  
8  c30d0232-10ee-45f0-9005-9270512c4abc  
9  3839234e-64c2-4e1d-91e4-458dc72dcfe8

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6376 entries, 0 to 6375
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   _id      6376 non-null   object
 1   Uid      6376 non-null   object
 2   name     6375 non-null   object
 3   Класс    6375 non-null   object
 4   Bay_Uid  6376 non-null   object
dtypes: object(5)
memory usage: 249.2+ KB


None

----------


_id ГруппПолномочий                Техместо  \
0  65a4daf1e20c332a8f15681b            754B  PS110-000002-12-01-001   
1  65a4daf1e20c332a8f15681c            754B  PS110-000002-12-01-002   
2  65a4daf1e20c332a8f15681d            754B  PS110-000002-12-01-201   
3  65a4daf1e20c332a8f15681e            754B  PS110-000002-12-01-202   
4  65a4daf1e20c332a8f15681f            754B  PS110-000002-12-01-330   
5  65a4daf1e20c332a8f156820            754B  PS110-000002-12-01-701   
6  65a4daf1e20c332a8f156821            754B  PS110-000002-14-01-006   
7  65a4daf1e20c332a8f156822            754B  PS110-000002-14-01-007   
8  65a4daf1e20c332a8f156823            754B  PS110-000002-14-01-008   
9  65a4daf1e20c332a8f156824            754B  PS110-000002-14-01-009   

                                   Название Дисп_наименование  Местоположение  \
0                                 ВЛ-110-50              None      7540020200   
1                                 ВЛ-110-84              None      7540020200   
2            Присоединение трансформатора 1              None      7540020200   
3            Присоединение трансформатора 2              None      7540020200   
4                        Строительная часть              None      7540020200   
5                Присоединение разрядника 1              None      7540020200   
6  Присоединение ВЛ-10кВ ф.Черемхово, яч.№6              None      7540020200   
7   Присоединение ВЛ-10 кВ ф.Хадакта, яч.№7              None      7540020200   
8   Присоединение ВЛ-10 кВ ф.Карымка, яч.№8              None      7540020200   
9               Присоединение 10 кВ ф.9 МТМ              None      7540020200   

  Наименование местоположения     ВерхнееТМ  Район    БЕ  ПланирующЗавод  \
0              Улетовский РЭС  PS110-000002    NaN  7500            7540   
1              Улетовский РЭС  PS110-000002    NaN  7500            7540   
2              Улетовский РЭС  PS110-000002    NaN  7500            7540   
3              Улетовский РЭС  PS110-000002    NaN  7500            7540   
4              Улетовский РЭС  PS110-000002    NaN  7500            7540   
5              Улетовский РЭС  PS110-000002    NaN  7500            7540   
6              Улетовский РЭС  PS110-000002    NaN  7500            7540   
7              Улетовский РЭС  PS110-000002    NaN  7500            7540   
8              Улетовский РЭС  PS110-000002    NaN  7500            7540   
9              Улетовский РЭС  PS110-000002    NaN  7500            7540   

   Видобъекта ОснСредство ПользовСтатус СистемнСтатус Вэксплуат_с  Годвыпуска  \
0        1023   T7500SRSK          ДСТВ          СОЗД         NaT         NaN   
1        1023   T7500SRSK          ДСТВ          СОЗД         NaT         NaN   
2        1023   T7500SRSK          ДСТВ          СОЗД         NaT         NaN   
3        1023   T7500SRSK          ДСТВ          СОЗД         NaT         NaN   
4        1023        None          ДСТВ          СОЗД         NaT         NaN   
5        1023   T7500SRSK          ДСТВ          СОЗД         NaT         NaN   
6        1023        None          ДСТВ          СОЗД         NaT         NaN   
7        1023        None          ДСТВ          СОЗД         NaT         NaN   
8        1023        None          ДСТВ          СОЗД         NaT         NaN   
9        1023        None          ДСТВ          СОЗД         NaT         NaN   

  Инвентарн№ НомерОСверх_ТМ  НомерОСинв_об_ Датасоздания        Создал  
0       None      T7500SRSK             NaN   2011-03-25  SHESTAKOV_EP  
1       None      T7500SRSK             NaN   2011-03-25  SHESTAKOV_EP  
2       None      T7500SRSK             NaN   2011-03-25  SHESTAKOV_EP  
3       None      T7500SRSK             NaN   2011-03-25  SHESTAKOV_EP  
4       None      T7500SRSK             NaN   2011-03-25  SHESTAKOV_EP  
5       None      T7500SRSK             NaN   2015-03-24  KAZACHENKOTA  
6       None      T7500SRSK             NaN   2011-03-25  SHESTAKOV_EP  
7       None      T7500SRSK          

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26453 entries, 0 to 26452
Data columns (total 22 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   _id                          26453 non-null  object        
 1   ГруппПолномочий              26453 non-null  object        
 2   Техместо                     26453 non-null  object        
 3   Название                     26453 non-null  object        
 4   Дисп_наименование            10259 non-null  object        
 5   Местоположение               26453 non-null  int64         
 6   Наименование местоположения  26453 non-null  object        
 7   ВерхнееТМ                    26453 non-null  object        
 8   Район                        0 non-null      float64       
 9   БЕ                           26453 non-null  int64         
 10  ПланирующЗавод               26453 non-null  int64         
 11  Видобъекта                   26453 non-nu

None

----------


_id                                   Uid  \
0  65a2362e2f6afc22094f7e23  3c6a054d-e274-464d-96bb-1400c71b3680   
1  65a2362e2f6afc22094f7e24  0c9156ce-53ee-4521-805f-a3870b1eeede   
2  65a2362e2f6afc22094f7e25  863922fe-3bce-4fac-8be2-756f22754403   
3  65a2362e2f6afc22094f7e26  dede43e7-ac05-4150-9e3f-6bb6a8b075f2   
4  65a2362e2f6afc22094f7e27  07b90e8a-8d13-47b7-8f83-8568eaf1dd9c   
5  65a2362e2f6afc22094f7e28  9460937d-42bd-493a-a047-42394c616ae8   
6  65a2362e2f6afc22094f7e29  1bde8b71-b9d4-43f8-8e12-11b0045031ed   
7  65a2362e2f6afc22094f7e2a  72751602-4f49-4f7d-8e51-90fc61152c23   
8  65a2362e2f6afc22094f7e2b  4dfb4bf9-d8cd-4293-8bfb-477307d83dcb   
9  65a2362e2f6afc22094f7e2c  61ba816e-1d4a-45c6-a083-9feb17698cba   

          Техместо Наименование Дисп_наименование         Класс  
0  PS110-000191-12          110         РУ-110 кВ  VoltageLevel  
1  PS110-000930-12          110         РУ-110 кВ  VoltageLevel  
2  PS110-000308-12          110         РУ-110 кВ  VoltageLevel  
3  PS110-000265-12          110         РУ-110 кВ  VoltageLevel  
4  PS110-000287-12          110         РУ-110 кВ  VoltageLevel  
5  PS110-000181-12          110         РУ-110 кВ  VoltageLevel  
6  PS110-000190-12          110         РУ-110 кВ  VoltageLevel  
7  PS110-000141-12          110         РУ-110 кВ  VoltageLevel  
8  PS110-000241-12          110         РУ-110 кВ  VoltageLevel  
9  PS110-000164-12          110         РУ-110 кВ  VoltageLevel

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 801 entries, 0 to 800
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   _id                801 non-null    object
 1   Uid                801 non-null    object
 2   Техместо           801 non-null    object
 3   Наименование       801 non-null    object
 4   Дисп_наименование  801 non-null    object
 5   Класс              801 non-null    object
dtypes: object(6)
memory usage: 37.7+ KB


None

----------


# Берем Техместо из РУ (ранее сопоставленных). Кладем в RS20

In [5]:
df_supa['Техместо_РУ'] = df_supa['Техместо'].apply(lambda x: x[:-7])

In [6]:
techplace_voltages_matched = set(df_voltages_matched['Техместо'].tolist())
techplace_voltages_supa = set(df_supa['Техместо_РУ'].tolist())
print('Между сопоставленными РУ и СУПА общих по Техместу РУ:', len(techplace_voltages_matched & techplace_voltages_supa))

uid_voltages_matched = set(df_voltages_matched['Uid'].tolist())
uid_voltages_rs = set(df_rs['Bay_Uid'].tolist())
print('Между сопоставленными РУ и RS20 общих по Uid РУ:', len(uid_voltages_matched & uid_voltages_rs))

Между сопоставленными РУ и СУПА общих по Техместу РУ: 728
Между сопоставленными РУ и RS20 общих по Uid РУ: 788


In [7]:
uid2techplace = dict(zip(df_voltages_matched['Uid'], df_voltages_matched['Техместо']))
test = []
for key in uid2techplace.keys():
    test.append(uid2techplace[key] == df_voltages_matched[df_voltages_matched['Uid'] == key]['Техместо'].values[0])

assert all(test) == True
df_rs['Техместо_РУ'] = df_rs['Bay_Uid'].apply(lambda x: uid2techplace.get(x))
print('None:', df_rs['Техместо_РУ'].isna().sum())
df_rs

None: 1736


_id                                   Uid    name  \
0     65a4dad6e20c332a8f154f33  c1413450-2c4e-4505-980b-ed4cc09737e4  РП-110   
1     65a4dad6e20c332a8f154f34  f9735b10-b60d-446a-b539-08b645e8a533  РП-110   
2     65a4dad6e20c332a8f154f35  cda400a2-f2db-4796-bb80-7bc2a9d652be  СП-110   
3     65a4dad6e20c332a8f154f36  167d5423-42aa-458e-90f9-4dc67013adea  СП-110   
4     65a4dad6e20c332a8f154f37  6a0433c6-f9b4-4d42-bb5e-4350177582e8  СП-110   
...                        ...                                   ...     ...   
6371  65a4dad6e20c332a8f156816  4744da0d-8729-44eb-a48a-06fa1bb9c15f  рп-110   
6372  65a4dad6e20c332a8f156817  af5d4fa8-0c6d-4b16-a720-63cfef6495c8  рп-110   
6373  65a4dad6e20c332a8f156818  022be731-f308-4f26-b987-745aabf15ebc  рп-110   
6374  65a4dad6e20c332a8f156819  830f7f5d-c5be-4ad5-8c00-7ba6737ccb60  РП-110   
6375  65a4dad6e20c332a8f15681a  3f1e7615-8aaa-440c-b764-006a2b20f74c  рп-110   

                                    Класс  \
0     Группа коммутационного оборудования   
1     Группа коммутационного оборудования   
2     Группа коммутационного оборудования   
3     Группа коммутационного оборудования   
4     Группа коммутационного оборудования   
...                                   ...   
6371  Группа коммутационного оборудования   
6372  Группа коммутационного оборудования   
6373  Группа коммутационного оборудования   
6374  Группа коммутационного оборудования   
6375  Группа коммутационного оборудования   

                                   Bay_Uid      Техместо_РУ  
0     f882ca3c-1f15-43ee-b604-834e430e3a2e  PS110-000276-12  
1     4d87d79c-3265-4689-8e93-aac597307e63  PS110-000202-12  
2     063e4acc-c24b-4ab6-b63a-e65bdc84f088  PS110-000271-12  
3     0ae198e5-68dd-409d-ad3a-c1ac6304002d  PS110-001210-12  
4     7bb39d2d-d23f-4c10-8807-34220c086532  PS110-000273-12  
...                                    ...              ...  
6371  a887ee57-dfd8-4c0c-9e17-bfd0148b9aee  PS110-000685-12  
6372  e83f370b-d483-41fd-8d01-f9799ee946ef  PS110-000606-12  
6373  27bfaf2f-7cbe-4c29-aacc-fcebc83157cb  PS110-000656-12  
6374  f640e82e-52ea-433a-880f-d08ad8e827a0  PS110-000590-12  
6375  0e8aa322-c3ae-4019-a0c4-25191d9e5808  PS110-000673-12  

[6376 rows x 6 columns]

In [8]:
texts = df_rs['name'].sample(50).tolist() + df_supa['Название'].sample(50).tolist() + df_supa['Дисп_наименование'].sample(50).tolist()

# Коды внутри названий
def find_code(text):
    if isinstance(text, str):
        match = re.findall('\(.*\)', text)
        if match:
            return match[0]\
                .replace('(', '')\
                .replace(')', '')\
                .replace(' ', '')\
                .replace('-', '')\
                .strip()
    return None

num2str = {
    '11': 'одиннадцать',
    '20': 'двадцать',
    '31': 'тридцать_один',
    '54': 'пятьдесят_четыре',
    '1': 'один',
    '2': 'два',
    '3': 'три',
    '4': 'четыре',
    '6': 'шесть',
}

def get_num(name, text):
    num = 0
    match = re.findall(f'{name} ?-? ?\d+', text)
    if match:
        num = re.findall('\d+', match[0])[0]
    text = re.sub(f'{name} ?№? ?-?\d', f'{name}-{num2str.get(num)}', text)
    return text

# Названия откуда - куда
def find_destination(text):
    if isinstance(text, str):
        match = re.findall('\(.*\)', text)
        if match:
            text = text[:text.index(match[0])]
            
        # text = get_num('Азот', text)
        # text = get_num('Сотая', text)
        # text = get_num('тяговая', text)
        # text = get_num('Тяговая', text)
        # text = get_num('Томусинская', text)
        # text = get_num('Товарищ', text)
        # text = get_num('Ширпотреб-Северная', text)
        # text = get_num('Ширпотреб-Прокатная', text)
        # text = get_num('Опорная', text)
        # text = get_num('Тырганская', text)
        # text = get_num('Абразивная', text)
        # text = get_num('Распадская 5', text)
        # text = get_num('Хвостохранилище', text)
        # text = get_num('Карьерная', text)
        # text = get_num('Разъезд', text)
        # text = get_num('ЯЦЗ', text)
        # text = get_num('Химпром', text)
        # text = get_num('Проскоковская', text)
        # text = get_num('Западно-Сибирская ТЭЦ', text)
        # text = get_num('Западно-Сибирская ТЭЦ - Опорная', text)
        # text = get_num('Ускатская-Луговая', text)
        # text = get_num('Темирская', text)
        # text = get_num('Шушталепская', text)
        # text = get_num('Красноярская', text)
                       
        text = re.sub('ё', 'е', text)
        text = re.sub('с отп.на', '', text)
        text = re.sub('с отп.', '', text)
        # text = re.sub('Цинзаводская', 'Цинкзаводская', text)
        # text = re.sub('Заискиимская', 'Заискитимская', text)
        # text = re.sub('Имена', 'Имени', text)
        text = re.sub('цепь', '', text)
        # text = re.sub('Беловская ГРЭС – Гурьевская ГРЭС', 'Беловская ГРЭС – Гурьевская', text)
        # text = re.sub('ЦентральнаяТЭЦ', 'Центральная ТЭЦ', text)
        # text = re.sub('СибирскаяТЭЦ', 'Сибирская ТЭЦ', text)
        # text = re.sub('[сc]{1}к\.', 'ский', text)
        # text = re.sub('Беллык', 'Беллыкский', text)
        # text = re.sub('[сc]{1}кая', 'ский', text)
        # text = re.sub('Беллыкскийский', 'Беллыкский', text)
        text = re.sub('К?ВЛ[ -]110 ?кВ', '', text)
        # text = re.sub('М.Имыш', 'Малый Имыш', text)
        # text = re.sub('К.Шалтырь', 'Кия-Шалтырь', text)
        # text = re.sub('Н.Огур', 'Новый Огур', text)
        # text = re.sub('Н.Сыда', 'Новая Сыда', text)
        # text = re.sub('Мусохраноская', 'Мусохрановская', text)
        text = re.sub('\d+ ?цепь', '', text)
        text = re.sub('[сc]? *отпай\w+ *(на)?', '', text)
        text = re.sub('[Н]?ПС', '', text)
        text = re.sub('ПП', '', text)
        # text = text.lower()
        # text = re.sub('обьект', 'объект', text)
        # text = re.sub('"', '', text)
        tokens = []
        g = tokenize(BytesIO(text.encode('utf-8')).readline)
        for toktype, tokval, _, _, _ in g:
            if toktype == NAME:
                tokens.append(tokval.strip())
        return '_'.join(tokens).strip()
    return None

def get_transformer_name(text):
    if isinstance(text, str):
        match = re.findall('[ТT]С?[-]?\d', text)
        if match:
            res = match[0].replace('-', '').replace('-', '').replace('T', 'Т').replace('С', '')
            return res
        match = re.findall('\d[ А]?Т', text)
        if match:
            res = match[0][-1]+match[0][0]
            return res
        match = re.findall('№\d', text)
        if match:
            num = re.findall('\d', match[0])
            return 'Т'+num[0]
    return None

def get_bay_num(text):
    if isinstance(text, str):
        match = re.findall('яч\.? ?№?\d+', text)
        if match:
            res = re.findall('\d+', match[0])
            if res:
                return res[0]
    return None
    
# for text in ['яч №7', 'яч.№17']:
#     print(text, '->', get_bay_num(text))

# for text in texts:
#     print(text, '->', get_transformer_name(text))
#     print(text, '->', find_code(text))
#     print(text, '->', find_destination(text))

001-190 Присоединение ЛЭП  
201-210 Присоединение трансформатора  
401-404 Присоединение ТН  
501-504 Присоединение ТСН  
610-611 Присоединение плавки гололеда  
601-602 Присоединение УКРМ  
701-704 Присоединение разрядника  
330 Строительная часть  

- Присоединение трансформатора PSNNN-123456-1+-0+-20+

    NNN класс напряжения ПС 110, 220, 330, 500 кВ,  
    XX класс напряжения РУ:  
    10 - 330  
    11 - 220  
    12 - 110  
     MM - Номер СШ 01-08, далее номер присоединения 20+, где + любое значение

- Присоединение ЛЭП(ПС) PSNNN-++++++-1+-0+-1++ 
- Присоединение ТН(ПС) PSNNN -++++++-1+-0+-40+ (Трансформа́тор напряже́ния)
- Присоединение ТСН(ПС) PSNNN-++++++-1+-0+-50+ (трансформаторы собственных нужд)
- Присоединение УКРМ PSNNN-++++++-1+-0+-60+ (УКРМ - Установки компенсации реактивной мощности)
- Присоединение плавки гололеда(ПС) PSNNN-++++++-1+-0+-61+ 
- Присоединение СВ, ШСВ(ПС) PSNNN-++++++-1+-11
- Присоединение СВ, ШСВ(ПС) PSNNN-++++++-1+-12
- Присоединение СВ, ШСВ(ПС) PSNNN-++++++-1+-13
- Присоединение СВ, ШСВ(ПС) PSNNN-++++++-1+-14
- Присоединение ОСВ(ПС) PSNNN-++++++-1+-21
- Присоединение ОСВ(ПС) PSNNN-++++++-1+-22
- Присоединение ОСВ(ПС) PSNNN-++++++-1+-23
- Присоединение ОСВ(ПС) PSNNN-++++++-1+-24
- Присоединение ОШСВ(ПС) PSNNN-++++++-1+-25
- Присоединение ОШСВ(ПС) PSNNN-++++++-1+-26
- Присоединение ОШСВ(ПС) PSNNN-++++++-1+-27
- Присоединение ОШСВ(ПС) PSNNN-++++++-1+-28
- Присоединение ремонтной перемычки(ПС) PSNNN-++++++-1+-29
- Присоединение ремонтной перемычки(ПС) PSNNN-++++++-1+-30
- Присоединение ремонтной перемычки(ПС) PSNNN-++++++-1+-31
- Присоединение ремонтной перемычки(ПС) PSNNN-++++++-1+-32
- Присоединение ДРГ PSNNN-++++++-01-80



In [9]:
df_rs['name'] = df_rs['name'].astype('str')
df_supa['Название'] = df_supa['Название'].astype('str')
df_supa['Дисп_наименование'] = df_supa['Дисп_наименование'].astype('str')
df_supa['type'] = df_supa['Техместо'].apply(lambda x: x[-3:])

In [10]:
def is_PowerTransformer(text):
    if isinstance(text, str):
        match = re.findall('^[ТT]-\d', text)
        if match:
            res = re.findall('\d', match[0])
            return res[0]
        match = re.findall('^\dТ', text)
        if match:
            res = re.findall('\d', match[0])
            return res[0]
    return None

df_rs['is_PowerTransformer'] = df_rs['name'].apply(is_PowerTransformer)
# df_rs[~df_rs['is_PowerTransformer'].isna()]['name'].unique()

In [22]:
# def is_Line_rs(text):
#     if isinstance(text, str):
#         if 'ВЛ-110-' in text:
#             match = re.findall('ВЛ-110-\d+', text)
#             if match:
#                 return match[0].replace(' ', '').replace('-', '').lower()
#         match = re.findall('^ВЛ', text.strip())
#         match = True
#         text = text.replace('ВЛ', '')
#         if match:
#             res1 = re.findall('[А-Я]{1,2} ?- ?\d+', text)
#             if res1:
#                 return res1[0].replace(' ', '').replace('-', '').lower()
#             res2 = re.findall('[а-яА-Я]+ ?- ?[а-яА-Я]+', text)
#             if res2:
#                 return res2[0].replace(' ', '').replace('-', '_').lower()
#     return None


def is_Line(text):
    if isinstance(text, str):
        if 'резерв' in text.lower():
            return None
        text1 = text.replace('ВЛ', '').replace('ЛР', '').replace('110', '').replace('II', '2').replace('I', '1')
        text1 = re.sub('ТЭЦ-\d', '', text1)
        match = re.findall('[А-Я]{1,2} *- *\d+', text1)
        if match:
            return match[0].replace(' ', '').replace('-', '').lower()
        match = re.findall('[а-яА-Я]+ *- *[а-яА-Я]+', text1)
        if match:
            res1 = match[0].replace(' ', '').replace('-', '_').lower()
            match1 = re.findall('\d цепь', text1)
            res2 = ''
            if match1:
                res2 = match1[0].replace('цепь', '').strip()
            return res1+res2
        #снова text
        if 'ВЛ-110-' in text:
            match = re.findall('ВЛ-110-\d+', text)
            if not match:
                match = re.findall('^ЛР', text)
            if match:
                return match[0].replace(' ', '').replace('-', '').lower()
        if 'ВЛ' in text or 'ЛР' in text:
            match = re.findall('[а-яА-Я]+ *- *[а-яА-Я]+', text)
            if not match:
                match1 = re.findall('[а-яА-Я]+', text)
                if match1:
                    return match1[0]
    return None

df_rs['is_Line'] = df_rs['name'].apply(is_Line)
# display(df_rs[~df_rs['is_Line'].isna()]['is_Line'].unique())

df_supa['is_Line'] = df_supa['Название'].apply(is_Line)
# df_supa['is_Line'] = df_supa['Дисп_наименование'].apply(is_Line)
# display(df_supa[~df_supa['is_Line'].isna()]['is_Line'].unique().tolist())

In [12]:
def is_Busbar(text):
    if isinstance(text, str):
        match1 = re.findall('с\.?ш\.?-? ?\d', text.lower())
        if match1:
            res = re.findall('\d', match1[0])
            return f'сш{res[0]}'
            
        match2 = re.findall('[\dО]сш', text.lower())
        if match2:
            res = re.findall('\d', match2[0])
            return f'сш{res[0]}'
    return None

df_rs['is_Busbar'] = df_rs['name'].apply(is_Busbar)
# display(df_rs[~df_rs['is_Busbar'].isna()]['name'].unique())
# display(df_rs[~df_rs['is_Busbar'].isna()]['is_Busbar'].unique())

df_supa['is_Busbar'] = df_supa['Название'].apply(is_Busbar)
# df_supa['is_Busbar'] = df_supa['Дисп_наименование'].apply(is_busbar)
# display(df_supa[~df_supa['is_Busbar'].isna()]['Название'].unique())

# merging

In [28]:
df_rs['PowerTransformer_key'] = df_rs['Техместо_РУ'] +'-0'+ df_rs['is_PowerTransformer'] + '-20x'
df_supa['PowerTransformer_key'] = df_supa['Техместо'].apply(lambda x: x[:-1]+'x')
feasible2match, df_merge1, df_rs, df_supa = merge_by_key_column(df_rs, df_supa, 'PowerTransformer_key')
feasible2match

63


0

In [23]:
df_rs['Line_key'] = df_rs['Техместо_РУ'] +'_'+ df_rs['is_Line']
df_supa['Line_key'] = df_supa['Техместо_РУ'] +'_'+ df_supa['is_Line']
feasible2match, df_merge2, df_rs, df_supa = merge_by_key_column(df_rs, df_supa, 'Line_key')
feasible2match

185


0

In [15]:
df_rs['Busbar_key'] = df_rs['Техместо_РУ'] +'_'+ df_rs['is_Busbar']
df_supa['Busbar_key'] = df_supa['Техместо_РУ'] +'_'+ df_supa['is_Busbar']
feasible2match, df_merge3, df_rs, df_supa = merge_by_key_column(df_rs, df_supa, 'Busbar_key')
feasible2match

37


27

In [24]:
df_supa['is_Line'] = df_supa['Дисп_наименование'].apply(is_Line)
df_supa['Line_key'] = df_supa['Техместо_РУ'] +'_'+ df_supa['is_Line']
feasible2match, df_merge4, df_rs, df_supa = merge_by_key_column(df_rs, df_supa, 'Line_key')
feasible2match

80


0

In [17]:
df_supa['is_Busbar'] = df_supa['Дисп_наименование'].apply(is_Busbar)
df_supa['Busbar_key'] = df_supa['Техместо_РУ'] +'_'+ df_supa['is_Busbar']
feasible2match, df_merge5, df_rs, df_supa = merge_by_key_column(df_rs, df_supa, 'Busbar_key')
feasible2match

4


0

In [25]:
feasible2match, df_merge6, df_rs, df_supa = merge_by_key_column(df_rs, df_supa, 'is_Line')
feasible2match

118


0

In [19]:
df_matched = pd.concat([df_merge1, df_merge2, df_merge3, df_merge4, df_merge5, df_merge6])
print('Сопоставлено', (df_matched.shape[0])/(df_supa.shape[0] + df_matched.shape[0]), 'СУПА')
print('Сопоставлено', (df_matched.shape[0])/(df_rs.shape[0] + df_matched.shape[0]), 'RS')

Сопоставлено 0.08241031262994745 СУПА
Сопоставлено 0.34190715181932246 RS


In [20]:
df_matched.shape, f'от прежних результатов: {df_matched.shape[0]/2935:.2}'

((2180, 45), 'от прежних результатов: 0.74')

In [21]:
df_rs['name'].unique().tolist()

['РП-110',
 'СП-110',
 'ЗОН-110 Т-1',
 'ЗОН Т-1',
 'ЗОН-110 Т-2',
 'ЗОН Т-2',
 'СР-110',
 'ТН-110-2',
 'Т-2',
 'НР-1Т',
 'Р-35-1',
 'Т-1',
 'ТН-110-1',
 'ЗН КС 110 Барнаульская ТЭЦ-3 - Власиха II цепь фаза «В»',
 'ЗОН-110 Т-3',
 'ОВ-110',
 'ТН 110-1',
 'СР',
 'ТН-110-3',
 'ВЛ СС-77',
 'ВЛ 110 кВ Корчинская - Мамонтовская (ВЛ КМ- 110)',
 'ЗОН-110 Т-4',
 'Т-2-40',
 'ТН 110-2',
 'ЗОН110 Т-2',
 'ВЛ 110 кВ Светлая-Буяновская (ВЛ СБ-427)',
 'СР-110-1',
 'СС-178',
 'СВ-110',
 'ДГР-35',
 'Т-1-40',
 'СР-110-2',
 'ЗОН-110-1Т',
 'ВЛ ГЗ-143 Горняк - Змеиногорская I цепь с отп.',
 'ВЛ СС-76',
 'ВЛ ЗС-31',
 'Присоединение Т-1',
 'Т-2-10',
 'ШСВ-110',
 'ЛР-110 Чегринская-Шебалинская',
 'ВЛ ГЗ-95 Горняцкая-Золотушинская',
 'ТН',
 'ВЛ ЮГ-154 Южная - Набережная с отп. на ПС Гидроузел',
 'ВЛ ПТ-62',
 'ВЛ ЧТ-181 Чергинская - Теньгинская с отп. на ПС Шебалинская',
 'ВЛ ЮБ-163 Южная - Безрукавская с отп. на ПС Алей',
 'ОД-110-1',
 'Т-3',
 'СП-110-2',
 'СП-110-1',
 'ВЛ-ГЗ-143',
 'ВЛ 110 кВ Бийская - Северо-З